sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientRetryPolicy.java

```diff
@@ -215,45 +236,81 @@ private Mono<ShouldRetryResult> shouldRetryOnStaleContainer() {
+        // own collection cache if they have one
         // Cosmos encryption is one use case
-        if(request.intendedCollectionRidPassedIntoSDK) {
-            return this.rxCollectionCache.refreshAsync(null, this.request).then( Mono.just(ShouldRetryResult.noRetry()));
+        if (request.intendedCollectionRidPassedIntoSDK) {
+            return this.rxCollectionCache.refreshAsync(null, this.request).then(Mono.just(ShouldRetryResult.noRetry()));
         }
 
-        //remove the previous header and try again
-        if(StringUtils.isNotEmpty(request.getHeaders().get(INTENDED_COLLECTION_RID_HEADER))) {
+        // remove the previous header and try again
+        if (StringUtils.isNotEmpty(request.getHeaders().get(INTENDED_COLLECTION_RID_HEADER))) {
             request.getHeaders().remove(INTENDED_COLLECTION_RID_HEADER);
         }
-        return this.rxCollectionCache.refreshAsync(null, this.request).then(Mono.just(ShouldRetryResult.retryAfter(Duration.ZERO)));
+        return this.rxCollectionCache.refreshAsync(null, this.request)
+                .then(Mono.just(ShouldRetryResult.retryAfter(Duration.ZERO)));
     }
 
-    private Mono<ShouldRetryResult> shouldRetryOnEndpointFailureAsync(boolean isReadRequest , boolean forceRefresh) {
+    private Mono<ShouldRetryResult> shouldRetryOnEndpointFailureAsync(boolean isReadRequest, boolean forceRefresh) {
         if (!this.enableEndpointDiscovery || this.failoverRetryCount > MaxRetryCount) {
             logger.warn("ShouldRetryOnEndpointFailureAsync() Not retrying. Retry count = {}", this.failoverRetryCount);
             return Mono.just(ShouldRetryResult.noRetry());
         }
 
         Mono<Void> refreshLocationCompletable = this.refreshLocation(isReadRequest, forceRefresh);
 
-        // Some requests may be in progress when the endpoint manager and client are closed.
+        // Some requests may be in progress when the endpoint manager and client are
+        // closed.
+        // In that case, the request won't succeed since the http client is closed.
+        // Therefore just skip the retry here to avoid the delay because retrying won't
+        // go through in the end.
+
+        Duration retryDelay = Duration.ZERO;
+        if (!isReadRequest) {
+            logger.debug("Failover happening. retryCount {}", this.failoverRetryCount);
+            if (this.failoverRetryCount > 1) {
+                // if retried both endpoints, follow regular retry interval.
+                retryDelay = Duration.ofMillis(ClientRetryPolicy.RetryIntervalInMS);
+            }
+        } else {
+            retryDelay = Duration.ofMillis(ClientRetryPolicy.RetryIntervalInMS);
+        }
+        return refreshLocationCompletable.then(Mono.just(ShouldRetryResult.retryAfter(retryDelay)));
+    }
+
+    private Mono<ShouldRetryResult> shouldRetryOnEndpointFailureAsync(boolean isReadRequest, boolean forceRefresh,
```

Comment:
I think we can combine the two `shouldRetryOnEndpointFailureAsync` methods? (the logicincluded are pretty similar)